<center>
  <font size="8">CS256 Team Project, 2023 Spring</font><br>
  <font size="6"> Team 2</font><br>
  Sai Sameer Yanamandra,Ganesh Regoti,Shanmukha Sri Harsha Anivilla,Nithish Reddy Agumamidi, Srujith Rao Ambati, Hemish Veeraboina
</center>

#### Essential info
- Machine learning algorithms used: Convoluted LSTM
- Bonus part(s): Part 1 and 2: and by whom (if any)
- Any other things ???

### Who did what
| What            | Description | By |
| ----------- | -----------       | ----------- |
| Resetting coordinate axes to nose, calculating polar angle, calculating polar velocity	| | Sameer, Ganesh, Nitish |
| Plots to graphs, answering questions, presentation	| | Hemish, Harsha, Srujith |
|Checking and verifying values for resetting coordinate axes to nose, checking and verifying values for polar angle, checking and verifying values for checking polar velocity, checking for other deviations | | 	Hemish, Srujith, Harsha|
|Checking for other deviations, plotting graphs, answering questions, presentation| | 	Sameer, Ganesh, Nitish|
|Splitting data, loading and calculating acceleration	| | Srujith, Nitish|
|Evaluating data, varying dropout, neurons and layers, summarizing scores	| | Harsha, Ganesh|
|Grid search, extrapolating video frames| | 	Hemish, Sameer|
| Presentation, verifying strategies for padding data, researching better techniques| | 	Srujith, Ganesh, Sameer|
|Testing train test split, validating dropout and layers,validating with our dataset| | 	Nitish, Hemish, Harsha|
|Testing train test split, validating dropout and layers, validating with our dataset for 3 gesture sequences| | 	Nitish, Hemish, Harsha|
|LSTM model for 3 gestures	| | Hemish, Sameer, Ganesh|
|Supporting functions for 3 gestures	| | Nitish, Srujith, Harsha|
|Video Segmentation| |  	Nitish, Srujith, Harsha|
|Implementation of sliding window and rms| | 	Hemish, Sameer, Ganesh|
|Testing 3 gestures	| | Nitish, Srujith, Harsha|
|Explaining results	| | Hemish, Sameer, Ganesh|
|  Bonus Item 1| | 	Nitish, Srujith, Harsha|
| Bonus Item 2| | 	Hemish, Sameer, Ganesh|


### Link to your online discussion forum for this project: [discussion link](http://???)

The following pair of gesture datasets are assigned to

(a) Warn, Turn left and turn right: Team #4

b) Warn, Turn  and circle : Team #3

c) Warn, Stop  and no : Team #2

d) Warn, Hello  and abort : Team #1

## We are working on classification of ???, ???, and ??? gestures

Please make sure
- Provide <font color= red>clear and explicit explanation</font> on the process and results in your work
- <font color= red>No ad hoc manual methodology</font> should be used in any task below.
- Provide <font color= red>generic solution</font> that is not specific/valid to your data file only
- Comments, in Markdown cells, should be added wherever appropriate
- Always use charts, figures, or other visual formats to present your results
- Copy and rename this file as teamXfinalproject.ipynb in your submission
- Your <font color= red>submitted Colab</font> can be <font color= red>run</font> from the beginning to the end <font color= red>without any error</font> ( you may opt for a different Colab file to showcase your bonus parts).
- Make sure all your <font color= red>Colab and data files are viewable to all SJSU</font> account holders.



## Project Result Outline
- [EDA and Data Preparation (selected milestone 1 content)](#EDA)
	- [Shift the origin of the data to the center of the body](#Shifttheorigin)
	- [Evaluating polar angle, angular velocity, and angular acceleration of from (x,y)](#angle-data)
	- [Store/write the polar angle velocity and polar angle acceleration data in the 3rd and the 4th component of each dataset.](#storing-angle-data)
- [Classification with LSTM (selected milestone 2 content)](#LSTM-classification)
	- [Data preparation for LSTM by Keras](#Keras-data)
	- [Split data to train(80%) and test(20%)](#Train-test-split)
	- [LSTM model performance with different configurations](#lstm-model)
	- [LSTM model performance with different datasets](#differnt-datasets)
	- [Reasons for the varying performance due to different model configurations or datasets](#performance-reasons)


- [classification of multiple gesture sequences(milestone 3)](#sequence)

	- [Creation of Concatenated videos with two gestures in a sequence](#videos)
	- [Creation of joint coordinate profiles of concatenated gestures](#joints)

	- [Modeling](#model)
	- [Discussion of your results using classification report](#classification_report)
	- [Explain your results](#explain)
  
  
- [Bonus item B1](#bonus-B1)
	- [9 gestures with 2-gesture sequence recognition](#9-2-gestures)

- [Bonus item B2](#bonus-B2)
	- [9 gestures with 3-gesture sequence recognition](#9-3-gestures)

- [Additional related results not shown above](#additional_results)

**Load packages:**

In [ ]:
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Everyone must put the copied folder in the exact path shown below so you collaborate and get graded easily.   
**No project score will be given for not following this folder configuration**    
(Our grader is *not reponsible to figure out your own* perferred Google *folder* configuration)   
**Only Keras or scikit_learn should be used in this project**

In [ ]:
root_path = "./drive/My Drive/CS256Project/deduced_data"

(os.path.exists(root_path)) #Checking if the data paths indeed exist and are valid.

True

- Import libraries and models needed for this work

In [ ]:
# lstm model
from numpy import mean
from numpy import std
from numpy import dstack
from pandas import read_csv
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Flatten
from keras.layers import Dropout
from keras.layers import LSTM
from keras.utils.np_utils import to_categorical
# from keras.utils import np_utils
from matplotlib import pyplot

In [ ]:
## your code here
import math

<a name="EDA"></a>
## EDA and Data Preparation (selected milestone 1 content)
---

In [ ]:
import colorsys
import os
from PIL import Image, ImageDraw

import matplotlib.pyplot as plt
import numpy as np

 You may use [this chrome app](https://script.google.com/macros/s/AKfycbxbGNGajrxv-HbX2sVY2OTu7yj9VvxlOMOeQblZFuq7rYm7uyo/exec) to copy the CS256Project folder that I shared with you to your own drive.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Everyone must put the copied folder in the exact path shown below so you collaborate and get graded easily.   
**No project score will be given for not following this folder configuration**    
(Our grader is *not reponsible to figure out your own* perferred Google *folder* configuration)   
Basically, you simply need to use the chrome app above to copy the entire CS256Project folder to your Google doc root folder

In [ ]:
gestures_data_path = "./drive/My Drive/CS256Project/data/"
gestures_2d_path = os.path.join(gestures_data_path, "gestures_basic_d2")

(os.path.exists(gestures_data_path),
 os.path.exists(gestures_2d_path)
) #Checking if the data paths indeed exist and are valid.

(True, True)

In [ ]:
# Sample Datapoint Properties
class_label = "abort"
video_id = "1_0"
n_frames = 24

## Exploring data

In [ ]:
# Extract a sample data and see what is in it
sample_2d_data_path = os.path.join(gestures_2d_path, class_label, f"{class_label}_{video_id}.mp4.npz")
data = np.load(sample_2d_data_path, allow_pickle=True)
# What does the data contain?
data.files

NameError: ignored

In [ ]:
data['keypoints'].shape

(24, 2)

### keypoints is where it stores the (x,y) coordindate values

In [ ]:
coords = data['keypoints']
coords[0,1].shape

(1, 4, 17)

### There are 17 labeled joints as generated from a research algorithm based on the human body image.

**The joint is indexed as follows**.   
0  - Nose   
1  - left eye   
2  - Right eye   
3  - left ear   
4  - Right ear   
5  - left shoulder   
6  - Right shoulder   
7  - left elbow   
8  - Right elbow   
9  - left wrist   
10  - Right wrist   
11  - left hip   
12  - Right hip   
13  - left knee   
14  - Right knee   
15  - left Ankle   
16  - Right Ankle   


<img alt="17 labeled joints" src="https://drive.google.com/uc?export=view&id=1fsCUlc7fetOIRpyL6_emdeFPzu1zJUHn">

### (x,y) coordinate of the first frame in the given sample

In [ ]:
coords[0][1][0][:,16]
# The first two components are (x,y) coordinate values

array([2.5641040e+02, 4.1656277e+02, 0.0000000e+00, 1.2290478e-02],
      dtype=float32)

### Let's take a tranponse so we can access the (x,y) coordindate easier

In [ ]:
# [0][1][0] because... [frame_id][always 1][0 to remove extra dimension]
# transpose(1, 0) to make it shape (17, 4) so first index is keypoint index
coords_0 = coords[0][1][0].transpose(1, 0)
coords_0.shape, coords_0

((17, 4),
 array([[2.9177484e+02, 1.8652383e+02, 0.0000000e+00, 1.1042018e+00],
        [3.0136520e+02, 1.7873605e+02, 0.0000000e+00, 1.5192554e+00],
        [2.8278391e+02, 1.7873605e+02, 0.0000000e+00, 1.7079070e+00],
        [3.1694955e+02, 1.8592477e+02, 0.0000000e+00, 1.2796223e+00],
        [2.7139536e+02, 1.8472665e+02, 0.0000000e+00, 1.4278147e+00],
        [3.4332303e+02, 2.4822697e+02, 0.0000000e+00, 2.5107551e-01],
        [2.4322366e+02, 2.4523167e+02, 0.0000000e+00, 2.3285715e-01],
        [3.4931699e+02, 3.2550568e+02, 0.0000000e+00, 4.2901537e-01],
        [2.3183511e+02, 3.2670383e+02, 0.0000000e+00, 4.9821147e-01],
        [3.5710916e+02, 4.0937405e+02, 0.0000000e+00, 4.2143461e-01],
        [2.2763934e+02, 4.0398254e+02, 0.0000000e+00, 3.7857029e-01],
        [3.2414230e+02, 4.0278439e+02, 0.0000000e+00, 7.1696095e-02],
        [2.6000681e+02, 3.9919003e+02, 0.0000000e+00, 1.0201875e-01],
        [3.2414230e+02, 4.7467157e+02, 0.0000000e+00, 1.6252953e-01],
        [2

<a name="Shifttheorigin"></a>
## Shift the origin to the nose and recalculate the new (x,y) coordinate for all datasets and
# overwrite the original (x,y) values

In [ ]:
cords=[]
for j in range(data['keypoints'].shape[0]):
  coords_0 = data['keypoints'][j][1][0].transpose(1, 0)
  nose_x,nose_y,_,__=coords_0[0]
  for i in range(len(coords_0)):
    coords_0[i][0]-=nose_x
    coords_0[i][1]-=nose_y
  cords.append(coords_0)

<a name="angle-data"></a>
# Calculate the polar angle velocity and the polar angle acceleration based on the transformed (x,y) from above

In [ ]:
from scipy.interpolate import interp1d
iter_d={"stop":{
    1:list(range(1,12)),
    2:list(range(0,9)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,11)),
    6:list(range(0,9)),

  },
   'no':{
    1:list(range(1,12)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,11)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'warn':{
    1:list(range(1,11)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   }
}


def polar_angle_velocity(class_label,video_id,n_frames):

  sample_2d_data_path = os.path.join(gestures_2d_path, class_label, f"{class_label}_{video_id}.mp4.npz")
  data = np.load(sample_2d_data_path, allow_pickle=True)
  # coords = data['keypoints']

  cords=[]
  for i in range(data['keypoints'].shape[0]):
    framei = data['keypoints'][i][1][0].transpose(1,0)
    n_x,n_y,_,_ = framei[0]
    framei[0][2]=framei[0][3]=0
    for j in range(1,len(framei)):

      framei[j][0]-=n_x
      framei[j][1]-=n_y
    cords.append(framei)

  for i in range(len(cords)):
    for j in range(len(cords[i])):
      cords[i][j][2] = math.atan2(cords[i][j][1],cords[i][j][0])
  for i in cords[0]:
    cords[0][3] = 0

  for i in range(1,len(cords)):
    prev=np.zeros_like(cords[0][0][2])
    for j in range(len(cords[i])):
      cords[i][j][3] = cords[i][j][2] - cords[i-1][j][2]
      cords[i][j][3] = cords[i][j][2] - prev
      prev = cords[i][j][2]
  cords = np.array(cords)
  if len(cords)<n_frames:
    # cords = np.moveaxis(cords, 0, -1)
    # num_features, num_datapoints, num_frames = cords.shape
    # # print(cords.shape)
    # # print(num_frames, cords)
    # f = interp1d(np.arange(num_frames), cords, axis=-1)
    # resampled_array = f(np.linspace(0, len(cords)-1, n_frames))

    # # Assign the resampled array back to the original array
    # cords = resampled_array
    # cords = np.moveaxis(cords, -1, 0)
    first_frame=cords[0]
    last_frame=cords[cords.shape[0]-1]
    n_=(n_frames-cords.shape[0])//2
    for i in range(n_):
      cords=np.insert(cords,0,first_frame, axis=0)
    for i in range(n_frames-cords.shape[0] ):
      cords=np.append(cords,np.array([last_frame]), axis=0)
  return cords

coords=[]
sizes={}
# coords=np.zeros_like(polar_angle_velocity("stop","1_1"))
for k in iter_d:
  v=iter_d[k]
  val=0
  for k_ in v:
    v_=v[k_]
    for i in v_:
      # print(k+f'{k_}_{i}')
      val+=1
      t=polar_angle_velocity(k,f'{k_}_{i}',63)
      coords.append(t)
  sizes[k]=val
      # coords=np.vstack((coords,polar_angle_velocity(k,f'{k_}_{i}')))
data=coords_np=np.array(coords)
print(coords_np.shape)




(182, 63, 17, 4)


In [ ]:
# np.concatenate(coords, axis=1).shape
# np.stack(coords).shape

In [ ]:
data.shape

(182, 63, 17, 4)

<a name="storing-angle-data"></a>
## Store the velocity and acceleration in the 3rd and 4th component of the dataset

In [ ]:
def load_acceleration(data):
    n_videos, n_frames, n_joints, n_data = data.shape
    array_with_acceleration = np.zeros((n_videos, n_frames, n_joints, n_data))

    for i, video in enumerate(data):
        # prev_frame=[]
        prev_frame = np.zeros_like(video[0, :, 3])
        for j, frame in enumerate(video):
            # if j==0:
            #   prev_frame=np.zeros_like(frame[:, 3])
            # polar_velocity = frame[:, 3]
            # print(polar_velocity)
            # polar_acceleration = calculate_polar_acceleration(polar_velocity)
            array_with_acceleration[i, j, :, 0] = frame[:, 0]  # x-coordinate
            array_with_acceleration[i, j, :, 1] = frame[:, 1]  # y-coordinate
            array_with_acceleration[i, j, :, 2] = frame[:, 3]  # polar velocity
            array_with_acceleration[i, j, :, 3] = frame[:, 3]-prev_frame
            # array_with_acceleration[i, j, :, 3] = polar_acceleration
            # print(frame[:, 3],prev_frame,frame[:, 3]-prev_frame)
            prev_frame=frame[:, 3]

    return array_with_acceleration
# data=load_acceleration(coords_np)


<a name="LSTM-classification"></a>
# Classification with LSTM (selected milestone 2 content)

<a name="Keras-data"></a>
## Data preparation for LSTM by Keras

<a name="Train-test-split"></a>
## Split data to train(80%) and test(20%)

In [ ]:
from sklearn.model_selection import train_test_split
print(data.shape)
print(sizes)
y_data= np.concatenate((np.array([1 for i in range(sizes['stop'])]),np.array([0 for i in range(sizes['no'])]),np.array([2 for i in range(sizes['warn'])])))
# y_data= np.concatenate((np.array([1 for i in range(sizes['stop'])]),np.array([0 for i in range(sizes['no'])])))
print(y_data.shape)
encoding={1:"stop",0:"no",2:'warn'}
X_train, X_test, y_train, y_test = train_test_split(coords_np, y_data, test_size=0.2, random_state=42, stratify=y_data)

(182, 63, 17, 4)
{'stop': 60, 'no': 62, 'warn': 60}
(182,)


In [ ]:
y_test.shape

(37,)

<a name="lstm-model"></a>
## LSTM model performance with different configurations

Here, we develop and evaluate a model to perform a classification task using Convolutional and Long Short-Term Memory (LSTM) networks.

Here's a summary of what the code does:

load_dataset function: This function accepts a series of inputs including attribute, training and testing datasets (features and labels). The function reshapes the data according to the requirement of the Convolutional and LSTM layers. It also customizes the reshaped data based on a given attribute, and converts the labels into categorical format using Keras' to_categorical function.

evaluate_model function: This function takes in the reshaped training and testing datasets, along with a few other parameters like the number of neurons in LSTM layers, dropout rate, number of LSTM layers, and the number of features. It reshapes the input data into 3D format which is necessary for Conv1D and LSTM layers.

The model architecture is as follows:

The first layer is a Conv1D layer that accepts time-series input. The number of filters and the kernel size for this layer are passed as arguments.

This is followed by a BatchNormalization layer, which normalizes the activations of the previous layer at each batch, i.e., applies a transformation that maintains the mean activation close to 0 and the activation standard deviation close to 1.

Then there are multiple LSTM layers, the number of which is controlled by the parameter n_lstm. If more than one LSTM layer is required, all LSTM layers except the last one will return sequences (3D output), making them suitable to be connected with the next LSTM layer.

The last LSTM layer does not return sequences and is followed by a Dropout layer (to prevent overfitting) and another BatchNormalization layer.

The final two layers are Dense layers, the first of which uses a 'relu' activation function and the second (the output layer) uses a 'softmax' activation function for multiclass classification. The number of neurons in the output layer is equal to the number of classes in the target variable.

The model is then compiled with 'categorical_crossentropy' as the loss function, 'adam' as the optimizer, and 'accuracy' as the evaluation metric.

In [ ]:
from keras.layers import Conv1D, TimeDistributed, Flatten,ConvLSTM2D
from keras.callbacks import EarlyStopping, ModelCheckpoint, ReduceLROnPlateau
from keras.models import Sequential
from keras.layers import Dense, Dropout, LSTM, Conv1D

# from keras_self_attention import SeqSelfAttention
from keras.layers import BatchNormalization
def load_dataset(attribute,trainX,testX,trainy,testy):
  # load all train
  print(trainX.shape)
  trainX = trainX.reshape(trainX.shape[0], 63, 17 , 4)
  # print(trainX[1][1])
  # trainX = load_acceleration(trainX)
  # print(trainX[1][1])
  print(f"loading {attribute}")
  l=list(range(5,11))
  print(l)
  # l.insert(0,0)
  if attribute=="velocity":
    trainX = trainX[:, :, l, 2:3]
  elif attribute=="acceleration":
    trainX = trainX[:, :, l, 3:]
  elif attribute=='both':
    trainX = trainX[:, :, l, 2:]
  elif attribute=='all':
    trainX = trainX[:, :, l, :]
  else:
    trainX = trainX[:, :, l, :]
  print(testX.shape)
  testX = testX.reshape(testX.shape[0], 63, 17 , 4)

  # testX = load_acceleration(testX)
  if attribute=="velocity":
    testX = testX[:, :, l, 2:3]
  elif attribute=="acceleration":
    testX = testX[:, :, l, 3:]
  elif attribute=='both':
    testX = testX[:, :, l, 2:]
  elif attribute=='all':
    testX = testX[:, :, l, :]
  else:
    testX = testX[:, :, l, :]
  trainy = to_categorical(trainy)
  testy = to_categorical(testy)
  print('[samples, time steps, features]',trainX.shape,  '[samples, time steps, features]',testX.shape)
  return trainX,testX,trainy,testy
def evaluate_model(trainX, trainy, testX, testy, neurons=50,dropout=0.5,n_lstm=2,fet=6,attr=1 ):
    verbose, epochs, batch_size = 0, 100, 32

    # Set the number of time steps and features according to your data
    time_steps = 63
    features = fet * attr
    n_conv_filters=64
    kernel_size=3

    # Reshape the input data to a 3D format
    trainX = trainX.reshape(trainX.shape[0], time_steps, features)
    testX = testX.reshape(testX.shape[0], time_steps, features)

    n_timesteps, n_features, n_outputs = trainX.shape[1], trainX.shape[2], trainy.shape[1]
    print(n_timesteps, n_features, n_outputs)
    model = Sequential()
    model.add(Conv1D(filters=n_conv_filters, kernel_size=kernel_size, activation='relu', input_shape=(n_timesteps, n_features)))
    model.add(Dropout(dropout))
    # model.add(BatchNormalization())


    if n_lstm>1:
      for i in range(n_lstm-1):
        model.add(LSTM(neurons, input_shape=(n_timesteps, n_features),return_sequences=True ))
        model.add(Dropout(dropout))
        # model.add(BatchNormalization())
    model.add( LSTM(int(1.5* neurons)) )
    model.add(Dropout(dropout))
    # model.add(BatchNormalization())
    model.add(Dense(300, activation='relu'))
    model.add(Dense(n_outputs, activation='softmax'))
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

    # fit network

    model.summary()
    # early_stopping = EarlyStopping(monitor='val_loss', patience=10)
    # model_checkpoint = ModelCheckpoint('best_model.h5', monitor='val_loss', save_best_only=True)
    # reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=0.001)
    # callbacks = [early_stopping, model_checkpoint, reduce_lr]

    model.fit(trainX, trainy, epochs=epochs, batch_size=batch_size, verbose=verbose)
    # probabilities = model.predict(testX, batch_size=batch_size)
    # probabilities = 0
    # evaluate model
    _,accuracy = model.evaluate(testX, testy, batch_size=batch_size, verbose=0)
    return model,accuracy


<a name="differnt-datasets"></a>
## LSTM model performance with different datasets

In [ ]:
# X_train, X_test, y_train, y_test
X_train_, X_test_, y_train_, y_test_ = load_dataset("acceleration", X_train, X_test,y_train, y_test,)
# trainX = np.load("./drive/MyDrive/CS256Project/deduced_data/data_temp/trainX.npz")["data"]
# trainy = np.load("./drive/MyDrive/CS256Project/deduced_data/data_temp/trainy.npz" )["data"]
# testX = np.load("./drive/MyDrive/CS256Project/deduced_data/data_temp/testX.npz")["data"]
# testy = np.load("./drive/MyDrive/CS256Project/deduced_data/data_temp/testy.npz")["data"]

# print(np.array_equal(X_train_,trainX))

print(X_train_.shape, X_test_.shape, y_train.shape, y_test.shape)
# trainX, trainy, testX, testy


(145, 63, 17, 4)
loading acceleration
[5, 6, 7, 8, 9, 10]
(37, 63, 17, 4)
[samples, time steps, features] (145, 63, 6, 1) [samples, time steps, features] (37, 63, 6, 1)
(145, 63, 6, 1) (37, 63, 6, 1) (145,) (37,)


Here, we are primarily concerned with the evaluation of the trained model on the test dataset and the generation of a classification report.

Here's a summary of what the code does:

The evaluate_model function is called with training and testing data along with parameters to define the model's architecture (number of LSTM layers, neurons per layer, dropout rate, and the number of features). This function returns a trained model and its accuracy on the test dataset, which are then printed out.

The trained model is used to make predictions on the test data using the model.predict method. The output of this prediction would be a probability distribution across classes, considering the last layer of the model is using softmax activation function.

These probabilities are then converted to class labels using the np.argmax function. The argmax function returns the index of the maximum value in each row (which is equivalent to the class label, given that the last layer is a softmax layer).

Similarly, if the test labels (y_test_) were one-hot encoded for training, they are also converted back to class labels.

A classification report is then generated using the classification_report function from sklearn.metrics. This report includes key classification metrics such as precision, recall, and f1-score for each class, as well as the overall accuracy of the model.

In [ ]:
from keras.layers import Reshape
from sklearn.metrics import classification_report
model,accuracy=evaluate_model(X_train_, y_train_,X_test_,  y_test_, neurons=150,dropout=0.2,n_lstm=2,fet=6,attr=1)
print(model,accuracy)
# model.re
predictions = model.predict(X_test_)

# If your output layer uses softmax activation, then predictions will be probabilities.
# You'll need to convert these probabilities into class labels. The argmax function can be used for this:
predictions_labels = np.argmax(predictions, axis=1)

# Similarly, if your testy is one-hot encoded, convert it back to labels
true_labels = np.argmax(y_test_, axis=1)

print(classification_report(true_labels, predictions_labels))
# model,accuracy2=evaluate_model(trainX, trainy,testX,  testy, neurons=150,dropout=0.2,n_lstm=2,fet=6,attr=1)
# print(model,accuracy2)

63 6 3
Model: "sequential_20"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv1d_21 (Conv1D)          (None, 61, 64)            1216      
                                                                 
 dropout_33 (Dropout)        (None, 61, 64)            0         
                                                                 
 lstm_44 (LSTM)              (None, 61, 150)           129000    
                                                                 
 dropout_34 (Dropout)        (None, 61, 150)           0         
                                                                 
 lstm_45 (LSTM)              (None, 225)               338400    
                                                                 
 dropout_35 (Dropout)        (None, 225)               0         
                                                                 
 dense_40 (Dense)            (None, 300)      

In [ ]:
print(accuracy)

0.6486486196517944


In [ ]:
import pickle
pickle.dump(model, open("./drive/MyDrive/CS256Project/data/data_for_lstm/finalized_model5.pickle", 'wb'))

<a name="performance-reasons"></a>
## Reasons for the varying performance due to different model configurations or datasets

The performance of our machine learning model can vary due to different model configurations and dataset properties.

Regarding model configuration, we designed a deep learning model with Convolutional and LSTM layers. Model hyperparameters, specifically the dropout rate and the number of neurons, play a critical role. We used grid search to determine optimal values for these parameters, settling on a dropout rate of 0.2 and 150 neurons.

The model's architecture and these hyperparameters significantly influence the model's learning capability and therefore its performance. An incorrect choice can lead to underfitting or overfitting of the model to the training data.

For the dataset, the size, quality, and complexity are factors that can affect the performance. It's important to have a large, clean, and balanced dataset that doesn't bias the model's learning process. For our case, we ensured a balanced representation of all labels in both training and testing sets by using stratified sampling during the train/test split.

In summary, both model configuration and dataset characteristics play crucial roles in model performance. Our approach involved careful selection of model hyperparameters and ensuring a balanced representation of the dataset for reliable model training and evaluation.

./drive/My Drive/CS256Project/data/gestures_basic_d2


<a name="optimal-model"></a>
# classification of multiple gesture sequences(milestone 3)

In [ ]:

def polar_angle_velocity_no_pad(class_label,video_id):

  sample_2d_data_path = os.path.join(gestures_2d_path, class_label, f"{class_label}_{video_id}.mp4.npz")
  data = np.load(sample_2d_data_path, allow_pickle=True)

  cords=[]
  for i in range(data['keypoints'].shape[0]):
    framei = data['keypoints'][i][1][0].transpose(1,0)

    cords.append(framei)

  cords = np.array(cords)

  return cords

<a name="GridSearchCV"></a>
## Creation of Concatenated videos with two gestures in a sequence
### All possible ordered cases out of your assigned 3 gestures must be created and saved in shared folders with matching names
---

In [ ]:
drive_path = "./drive/MyDrive/CS256Project/deduced_data/"
iter_d={"stop":{
    1:list(range(1,12)),
    2:list(range(0,9)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,11)),
    6:list(range(0,9)),

  },
   'no':{
    1:list(range(1,12)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,11)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'warn':{
    1:list(range(1,11)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   }
}
for k in iter_d:
  v=iter_d[k]
  # val=0
  for k_ in v:
    v_=v[k_]
    # print(f"for {k} - {k_} - {v_} ")
    for i in v_:
      # print(k+f'{k_}_{i}')
      # val+=1
      t=polar_angle_velocity_no_pad(k,f'{k_}_{i}')
      for k2 in iter_d:
        if k2==k:
          continue
        v2=iter_d[k2]
        # val=0
        for k_2 in v2:
          v_2=v2[k_2]
          # print(v_2)
          # print(f"for 2 {k2} - {k_2} - {v_2} ")
          for i2 in v_2:
            t2=polar_angle_velocity_no_pad(k2,f'{k_2}_{i2}')
            t3=np.concatenate((t,t2))
            # print(f'{k}_{k_}_{i}_{k2}_{k_2}_{i2}.npz')
            if not os.path.exists(drive_path+f"{k}_{k2}/"):
              os.mkdir(os.path.join(drive_path,f"{k}_{k2}"))
            np.savez(drive_path+f"{k}_{k2}/"+f'{k}_{k_}_{i}_{k2}_{k_2}_{i2}.npz',data=t3)
    print(f"{k} {k_}")
      # coords=np.vstack((coords,polar_angle_velocity(k,f'{k_}_{i}')))
# coords_np=np.array(coords)

stop 1
stop 2
stop 3
stop 4
stop 5
stop 6
no 1
no 2
no 3
no 4
no 5
no 6
warn 1
warn 2
warn 3
warn 4
warn 5
warn 6


<a name="bconfusion matrix and classification report for analyzing the binary classification result"></a>
## Creation of joint coordinate profiles of concatenated gestures
---

<a name="baseline-accuracy"></a>
## Modeling
---

The get_window function plays a critical role in the identification and extraction of informative segments from the sequence data. The fundamental principle here is to isolate windows where the start and end frames exhibit minimal differences, as quantified by the Euclidean distance between them. This method is effective in targeting sections of data where distinct yet similar patterns exist. The data within these windows can then be examined further or used to feed into a machine learning model for tasks like anomaly detection, event prediction, or behavior recognition.

These identified windows, however, might be too closely spaced and might not offer distinct information individually. To ensure that the selected windows are meaningfully spaced apart, we use the rms function. It examines the list of ending indices of windows and filters out those which are less than 17 units apart from the previous one. This step ensures that we are not over-representing very similar data in subsequent analyses or modeling. By applying this method, we obtain a series of data windows that are not only representative of unique patterns within the sequence but also spaced at a reasonable distance from each other. This can significantly help improve the robustness and generalizability of any models trained on this data.

The provided code illustrates an innovative approach to video segmentation using the concept of Euclidean distance. Euclidean distance is a measure of the straight-line distance between two points in a space. In this context, it's used to determine the similarity between different frames in a video, enabling us to segment the video into distinct clips.

Window Selection: In the get_window function, a window of frames is slid through each video. For each possible start frame within a video, it computes the Euclidean distance between the start frame and each frame after the minimum length away. The frame with the minimum distance (most similar) to the start frame is chosen as the end frame for that segment, and this process repeats starting from the frame after the selected end frame until the entire video is segmented.

This approach for video segmentation is especially advantageous as it ensures the selection of distinct and representative frames for each segment. By utilizing Euclidean distance, it can identify the start and end frames that are most similar over a minimum length of time, thereby capturing segments of the video where less change happens. This method allows for more coherent and meaningful video segmentation, which can be particularly beneficial in tasks such as action recognition where capturing complete actions within each segment is crucial.



In [ ]:
from scipy.spatial import distance

def polar_velocity_acceleration_(data):

  cords=[]
  # print(data.shape)
  for i in range(data.shape[0]):
    # print(data[i].shape)
    # data[i] = data[i][0]
    # print(data[i])
    n_x,n_y,_,_ = data[i][0]
    data[i][2]=data[i][3]=0
    for j in range(0,len(data[i])):

      data[i][j][0]-=n_x
      data[i][j][1]-=n_y
    cords.append(data[i])

  for i in range(len(cords)):
    for j in range(len(cords[i])):
      cords[i][j][2] = math.atan2(cords[i][j][1],cords[i][j][0])
  for i in cords[0]:
    cords[0][3] = 0

  for i in range(1,len(cords)):
    prev=np.zeros_like(cords[0][0][2])
    for j in range(len(cords[i])):
      cords[i][j][3] = cords[i][j][2] - cords[i-1][j][2]
      cords[i][j][3] = cords[i][j][2] - prev
      prev = cords[i][j][2]
  cords = np.array(cords)

  return cords


def transpose_data(data_):
  data=[]
  # print(data_.shape)
  for j in range(len(data_)):
    coords_0 = data_[j][1][0].transpose(1, 0)
    data.append(coords_0)

  data=np.array(data)
  return data

def pad_(cords,n_frames):
  if cords.shape[0]<n_frames:
    first_frame=cords[0]
    last_frame=cords[cords.shape[0]-1]
    n_=(n_frames-cords.shape[0])//2
    for i in range(n_):
      cords=np.insert(cords,0,first_frame, axis=0)
    for i in range(n_frames-cords.shape[0] ):
      cords=np.append(cords,np.array([last_frame]), axis=0)
  return cords

def get_window(data):
  min_length = 18
  # data=transpose_data(data)
  # frames = data.shape
  sfno = 0
  # min_dists = []
  # print(data.shape)
  min_indices = []
  while sfno < len(data) - min_length:
      dist_l = []
      # print(len(dist_l), sfno, len(data), min_length)
      for i in range(sfno, len(data) - min_length):
          start_frame = data[sfno,  5:11, :2].reshape(-1)
          end_frame = data[i + min_length, 5:11, :2].reshape(-1)

          dist = distance.euclidean(start_frame, end_frame)
          dist_l.append(dist)
      # print(len(dist_l), sfno, len(data), min_length,len(dist_l)-5)
      # print(dist_l)
      # print(dist_l[:])
      min_dist=None
      if len(dist_l) <=11:
        min_dist = min(dist_l[:])
      elif sfno==0:
        min_dist = min(dist_l[:len(dist_l)-10])
      else:
        min_dist = min(dist_l[5:])
      min_dist_index = dist_l.index(min_dist)

      # min_dists.append(min_dist)
      min_indices.append(sfno + min_dist_index + 18)

      sfno += min_dist_index + 3 + 18
      # print(dist_l)
  # min_indices.append(frames+1)
  return min_indices
# print(data.shape)
# print(get_window(data))

def rms(lst):
    i = 0
    while i < len(lst) - 1:
        if abs(lst[i] - lst[i + 1]) < 17:
            del lst[i + 1]
        else:
            i += 1
    return lst


# min_indices=get_window(data)
# min_indices = rms(min_indices)

In [ ]:
import numpy as np

# Load the data
data1 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/stop/stop_6_1.mp4.npz', allow_pickle=True)
data2 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_2_1.mp4.npz', allow_pickle=True)
# data3 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_1_5.mp4.npz', allow_pickle=True)

print(data1["keypoints"].shape)
print(data2["keypoints"].shape)
# print(data3["keypoints"].shape)
# Concatenate the arrays along the first axis (frames)
# video_concat = np.concatenate((data1['keypoints'], data2['keypoints'],data3['keypoints']), axis=0)
video_concat = np.concatenate((data1['keypoints'], data2['keypoints']), axis=0)

# Print the shape of the concatenated video
print(f'Shape of concatenated video: {video_concat.shape}')





(23, 2)
(27, 2)
Shape of concatenated video: (50, 2)


In [ ]:
drive_path = "./drive/MyDrive/CS256Project/deduced_data/"
def test(model,label1,label2,vid1,vid2):
  data = np.load(f"{drive_path}{label1}_{label2}/{label1}_{vid1}_{label2}_{vid2}.npz", allow_pickle=True)
  # data2 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_2_1.mp4.npz', allow_pickle=True)
  data=data["data"]
  # print(data.shape)
  min_indices=get_window(data)
  # print(min_indices)
  min_indices = rms(min_indices)
  # print(min_indices)
  data=polar_velocity_acceleration_(data)
  t=pad_(data[:min_indices[0]+1],63)[0:63,5:11 ,3:]
  t=np.array([t])
  # print(t.shape)
  t=t.reshape(t.shape[0], 63, 6)
  # print(t.shape)
  # print()
  p1=encoding[np.argmax(model.predict(t))]

  t1=pad_(data[min_indices[0]+3:],63)[0:63,5:11 ,3:]
  t1=np.array([t1])
  t1=t1.reshape(t1.shape[0], 63, 6)
  p2=encoding[np.argmax(model.predict(t1))]

  # print(f"label vs pred 1-> {label1==p1} for :{label1},{p1}")
  # print(f"label vs pred 2-> {label2==p2} for :{label2},{p2}")
  return label1==p1 and label2==p2




In [ ]:
data = np.load(f"{drive_path}no_stop/no_1_1_stop_3_5.npz", allow_pickle=True)
# data2 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_2_1.mp4.npz', allow_pickle=True)
data=data["data"]
data.shape

(55, 17, 4)

In [ ]:
# for i in iter_d:
t=[]
for j in iter_d["warn"]:
  for k in iter_d["warn"][j]:
    t.append(test(model,"warn","stop",f"{j}_{k}","3_5"))
    print("warn_"+f"{j}_{k}_"+"stop_"+f"{k%6}_{(j%8)}_"+f"{t[-1]}" )


1/1 [==============================] - 0s 37ms/step
warn_1_1_stop_1_1_True
1/1 [==============================] - 0s 36ms/step
warn_1_2_stop_2_1_True
1/1 [==============================] - 0s 36ms/step
warn_1_3_stop_3_1_True
1/1 [==============================] - 0s 36ms/step
warn_1_4_stop_4_1_True
1/1 [==============================] - 0s 46ms/step
warn_1_5_stop_5_1_True
1/1 [==============================] - 0s 66ms/step
warn_1_6_stop_0_1_True
1/1 [==============================] - 0s 36ms/step
warn_1_7_stop_1_1_True
1/1 [==============================] - 0s 37ms/step
warn_1_8_stop_2_1_True
1/1 [==============================] - 0s 35ms/step
warn_1_9_stop_3_1_True
1/1 [==============================] - 0s 50ms/step
warn_1_10_stop_4_1_True
1/1 [==============================] - 0s 43ms/step
warn_2_0_stop_0_2_True
1/1 [==============================] - 0s 36ms/step
warn_2_1_stop_1_2_True
1/1 [==============================] - 0s 66ms/step
warn_2_2_stop_2_2_True
1/1 [===================

In [ ]:
t=[]
for i in ["no",'warn']:
  for j in iter_d[i]:
    for k in iter_d[i][j]:
      t.append(test(model,i,"stop",f"{j}_{k}",f"{(k%5)+1}_{(j%8)+1}"))
      print(i+"_"+f"{j}_{k}_"+"stop_"+f"{(k%5)+1}_{(j%8)+1}_"+f"{t[-1]}" )
for i in ["stop",'warn']:
  for j in iter_d[i]:
    for k in iter_d[i][j]:
      t.append(test(model,i,"no",f"{j}_{k}",f"{(k%5)+1}_{(j%8)+1}"))
      print(i+"_"+f"{j}_{k}_"+"no_"+f"{(k%5)+1}_{(j%8)+1}_"+f"{t[-1]}" )
for i in ["no",'stop']:
  for j in iter_d[i]:
    for k in iter_d[i][j]:
      t.append(test(model,i,"warn",f"{j}_{k}",f"{(k%5)+1}_{(j%8)+1}"))
      print(i+"_"+f"{j}_{k}_"+"warn_"+f"{(k%5)+1}_{(j%8)+1}_"+f"{t[-1]}" )



1/1 [==============================] - 0s 37ms/step
no_1_1_stop_2_2_True
1/1 [==============================] - 0s 41ms/step
no_1_2_stop_3_2_False
1/1 [==============================] - 0s 41ms/step
no_1_3_stop_4_2_False
1/1 [==============================] - 0s 39ms/step
no_1_4_stop_5_2_False
1/1 [==============================] - 0s 39ms/step
no_1_5_stop_1_2_True
1/1 [==============================] - 0s 46ms/step
no_1_6_stop_2_2_True
1/1 [==============================] - 0s 38ms/step
no_1_7_stop_3_2_True
1/1 [==============================] - 0s 36ms/step
no_1_8_stop_4_2_False
1/1 [==============================] - 0s 39ms/step
no_1_9_stop_5_2_False
1/1 [==============================] - 0s 33ms/step
no_1_10_stop_1_2_True
1/1 [==============================] - 0s 41ms/step
no_1_11_stop_2_2_False
1/1 [==============================] - 0s 38ms/step
no_2_0_stop_1_3_True
1/1 [==============================] - 0s 36ms/step
no_2_1_stop_2_3_False
1/1 [==============================] - 0s 

In [ ]:
from sklearn.metrics import classification_report
print(f"True Count",t.count(True),f"|| Total Number ",len(t),f"|| Accuracy :",(t.count(True)/len(t))*100)
y_true=[True for i in range(len(t))]
print(classification_report(y_true, t))


True Count 181 || Total Number  364 || Accuracy : 49.72527472527473
              precision    recall  f1-score   support

       False       0.00      0.00      0.00         0
        True       1.00      0.50      0.66       364

    accuracy                           0.50       364
   macro avg       0.50      0.25      0.33       364
weighted avg       1.00      0.50      0.66       364



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


<a name="CNN-LSTM-model"></a>
## Discussion of your results using classification report
---

<a name="factor-summary"></a>
## Explain your results
---

In our experiment, we achieved an accuracy of 49.7%. However, some factors could have potentially affected the performance of our model leading to this accuracy score.

Firstly, inconsistencies in the data could be a significant factor. If there is a high degree of variability or noise in the data, the model may have difficulty identifying the relevant patterns, which could potentially lower its prediction accuracy.

Additionally, the number of frames in each video could also influence the model's performance. Videos with fewer frames provide less temporal information, and as a result, it can be more challenging for the model to capture the necessary sequential patterns. The lack of sufficient data could, therefore, result in a lower prediction accuracy.

Moreover, inconsistencies in the frames, such as abrupt changes, noise, or missing information, can disrupt the patterns that the model is trying to learn. If the frames are not consistent, the model may end up learning incorrect patterns, which could lead to lower accuracy in its predictions.

In essence, the quality and quantity of the video data can significantly affect the model's ability to learn, and thus its final performance. Therefore, it is crucial to ensure that the data is as clean and consistent as possible and that there are enough frames in the video for the model to capture relevant patterns effectively.

<a name="bonus-B1"></a>
# Bonus item B1
---

<a name="4-gestures"></a>
## Extend your work to training a model with all 9 gestures recognition for 2-gesture sequence recognition
---

In [ ]:
iter_d={"stop":{
    1:list(range(1,12)),
    2:list(range(0,9)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,11)),
    6:list(range(0,9)),

  },
   'no':{
    1:list(range(1,12)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,11)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'warn':{
    1:list(range(1,11)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'circle':{
    1:list(range(1,10)),
    2:list(range(0,11)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'turn':{
    1:list(range(1,10)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,11)),
    6:list(range(0,13)),

   },
   'turn_left':{
    1:list(range(1,12)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,11)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'turn_right':{
    1:list(range(1,10)),
    2:list(range(0,11)),
    3:list(range(0,9)),
    4:list(range(0,10)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'hello':{
    1:list(range(1,11)),
    2:list(range(0,10)),
    3:list(range(0,10)),
    4:list(range(0,9)),
    5:list(range(0,10)),
    6:list(range(0,10)),

   },
   'abort':{
    1:list(range(0,10)),
    2:list(range(0,9)),
    3:list(range(0,10)),
    4:list(range(0,10)),
    5:list(range(0,9)),
    6:list(range(0,9)),

   }

}
l=[]
coords=[]
for k in iter_d:
  l.append(k)
  v=iter_d[k]
  val=0
  for k_ in v:
    v_=v[k_]
    for i in v_:
      # print(k+f'{k_}_{i}')
      val+=1
      t=polar_angle_velocity(k,f'{k_}_{i}',120)
      coords.append(t)
  sizes[k]=val
      # coords=np.vstack((coords,polar_angle_velocity(k,f'{k_}_{i}')))
data=coords_np=np.array(coords)
print(coords_np.shape)

(542, 120, 17, 4)


In [ ]:
# data=transpose_data(coords_np)
data.shape

(542, 120, 17, 4)

In [ ]:
print(data.shape)
print(sizes)
print(l)
# for i in range(0,9):
y_data= np.concatenate((np.array([1 for i in range(sizes['stop'])]),np.array([0 for i in range(sizes['no'])]),np.array([2 for i in range(sizes['warn'])]),np.array([3 for i in range(sizes['circle'])]),np.array([4 for i in range(sizes['turn'])]),np.array([5 for i in range(sizes['turn_left'])]),np.array([6 for i in range(sizes['turn_right'])]),np.array([7 for i in range(sizes['hello'])]),np.array([8 for i in range(sizes['abort'])])))
# y_data= np.concatenate((np.array([1 for i in range(sizes['stop'])]),np.array([0 for i in range(sizes['no'])])))
print(y_data.shape)

encoding={1:"stop",0:"no",2:'warn',3:"circle",4:"turn",5:"turn_left",6:"turn_right",7:"hello",8:"abort"}
X_train, X_test, y_train, y_test = train_test_split(coords_np, y_data, test_size=0.2, random_state=42, stratify=y_data)

(542, 120, 17, 4)
{'stop': 60, 'no': 62, 'warn': 60, 'circle': 60, 'turn': 63, 'turn_left': 62, 'turn_right': 59, 'hello': 59, 'abort': 57}
['stop', 'no', 'warn', 'circle', 'turn', 'turn_left', 'turn_right', 'hello', 'abort']
(542,)


In [ ]:
from keras.layers import Conv1D, TimeDistributed, Flatten,ConvLSTM2D
from keras.callbacks import EarlyStopping, ModelCheckpoint, ReduceLROnPlateau
from keras.models import Sequential
from keras.layers import Dense, Dropout, LSTM, Conv1D

# from keras_self_attention import SeqSelfAttention
from keras.layers import BatchNormalization
def load_dataset_2(attribute,trainX,testX,trainy,testy,):
  # load all train
  print(trainX.shape)
  trainX = trainX.reshape(trainX.shape[0], 120, 17 , 4)
  # print(trainX[1][1])
  # trainX = load_acceleration(trainX)
  # print(trainX[1][1])
  print(f"loading {attribute}")
  l=list(range(5,11))
  print(l)
  # l.insert(0,0)
  if attribute=="velocity":
    trainX = trainX[:, :, l, 2:3]
  elif attribute=="acceleration":
    trainX = trainX[:, :, l, 3:]
  elif attribute=='both':
    trainX = trainX[:, :, l, 2:]
  elif attribute=='all':
    trainX = trainX[:, :, l, :]
  else:
    trainX = trainX[:, :, l, :]
  print(testX.shape)
  testX = testX.reshape(testX.shape[0], 120, 17 , 4)

  # testX = load_acceleration(testX)
  if attribute=="velocity":
    testX = testX[:, :, l, 2:3]
  elif attribute=="acceleration":
    testX = testX[:, :, l, 3:]
  elif attribute=='both':
    testX = testX[:, :, l, 2:]
  elif attribute=='all':
    testX = testX[:, :, l, :]
  else:
    testX = testX[:, :, l, :]
  trainy = to_categorical(trainy)
  testy = to_categorical(testy)
  print('[samples, time steps, features]',trainX.shape,  '[samples, time steps, features]',testX.shape)
  return trainX,testX,trainy,testy
def evaluate_model_2(trainX, trainy, testX, testy, neurons=50,dropout=0.5,n_lstm=2,fet=6,attr=1 ):
    verbose, epochs, batch_size = 0, 100, 32

    # Set the number of time steps and features according to your data
    time_steps = 120
    features = fet * attr
    n_conv_filters=64
    kernel_size=3

    # Reshape the input data to a 3D format
    trainX = trainX.reshape(trainX.shape[0], time_steps, features)
    testX = testX.reshape(testX.shape[0], time_steps, features)

    n_timesteps, n_features, n_outputs = trainX.shape[1], trainX.shape[2], trainy.shape[1]
    print(n_timesteps, n_features, n_outputs)
    model = Sequential()
    model.add(Conv1D(filters=n_conv_filters, kernel_size=kernel_size, activation='relu', input_shape=(n_timesteps, n_features)))
    model.add(BatchNormalization())
    model.add(Dropout(dropout))


    if n_lstm>1:
      for i in range(n_lstm-1):
        model.add(LSTM(neurons, input_shape=(n_timesteps, n_features),return_sequences=True ))
        model.add(BatchNormalization())
        model.add(Dropout(dropout))
    model.add( LSTM(int(1.5* neurons)) )
    model.add(BatchNormalization())
    model.add(Dropout(dropout))
    model.add(Dense(300, activation='relu'))
    model.add(Dense(n_outputs, activation='softmax'))
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

    # fit network

    model.summary()
    # early_stopping = EarlyStopping(monitor='val_loss', patience=10)
    # model_checkpoint = ModelCheckpoint('best_model.h5', monitor='val_loss', save_best_only=True)
    # reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=0.001)
    # callbacks = [early_stopping, model_checkpoint, reduce_lr]

    model.fit(trainX, trainy, epochs=epochs, batch_size=batch_size, verbose=verbose)
    # probabilities = model.predict(testX, batch_size=batch_size)
    # probabilities = 0
    # evaluate model
    _,accuracy = model.evaluate(testX, testy, batch_size=batch_size, verbose=0)
    return model,accuracy


In [ ]:
X_train_, X_test_, y_train_, y_test_ = load_dataset_2("acceleration", X_train, X_test,y_train, y_test,)

(433, 120, 17, 4)
loading acceleration
[5, 6, 7, 8, 9, 10]
(109, 120, 17, 4)
[samples, time steps, features] (433, 120, 6, 1) [samples, time steps, features] (109, 120, 6, 1)


In [ ]:
from keras.layers import Reshape
from sklearn.metrics import classification_report
model,accuracy=evaluate_model_2(X_train_, y_train_,X_test_,  y_test_, neurons=200,dropout=0.3,n_lstm=2,fet=6,attr=1)
print(model,accuracy)
predictions = model.predict(X_test_)

predictions_labels = np.argmax(predictions, axis=1)

true_labels = np.argmax(y_test_, axis=1)

print(classification_report(true_labels, predictions_labels))

120 6 9
Model: "sequential_22"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv1d_23 (Conv1D)          (None, 118, 64)           1216      
                                                                 
 batch_normalization_51 (Bat  (None, 118, 64)          256       
 chNormalization)                                                
                                                                 
 dropout_39 (Dropout)        (None, 118, 64)           0         
                                                                 
 lstm_48 (LSTM)              (None, 118, 200)          212000    
                                                                 
 batch_normalization_52 (Bat  (None, 118, 200)         800       
 chNormalization)                                                
                                                                 
 dropout_40 (Dropout)        (None, 118, 200)

In [ ]:
drive_path = "./drive/MyDrive/CS256Project/deduced_data/"
def test(model,label1,label2,vid1,vid2):
  data = np.load(f"{drive_path}{label1}_{label2}/{label1}_{vid1}_{label2}_{vid2}.npz", allow_pickle=True)
  # data2 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_2_1.mp4.npz', allow_pickle=True)
  data=data["data"]
  # print(data.shape)
  min_indices=get_window(data)
  # print(min_indices)
  min_indices = rms(min_indices)
  # print(min_indices)
  data=polar_velocity_acceleration_(data)
  t=pad_(data[:min_indices[0]+1],63)[0:63,5:11 ,3:]
  t=np.array([t])
  # print(t.shape)
  t=t.reshape(t.shape[0], 63, 6)
  # print(t.shape)
  # print()
  p1=encoding[np.argmax(model.predict(t))]

  t1=pad_(data[min_indices[0]+3:],63)[0:63,5:11 ,3:]
  t1=np.array([t1])
  t1=t1.reshape(t1.shape[0], 63, 6)
  p2=encoding[np.argmax(model.predict(t1))]

  print(f"label vs pred 1-> {label1==p1} for :{label1},{p1}")
  print(f"label vs pred 2-> {label2==p2} for :{label2},{p2}")
  return label1==p1 and label2==p2




In [ ]:
import numpy as np
def test2(model,label1,label2,vid1,vid2):
  # Load the data
  data1 = np.load(f'./drive/My Drive/CS256Project/data/gestures_basic_d2/{label1}/{label1}_{vid1}.mp4.npz', allow_pickle=True)['keypoints']
  data2 = np.load(f'./drive/My Drive/CS256Project/data/gestures_basic_d2/{label2}/{label2}_{vid2}.mp4.npz', allow_pickle=True)['keypoints']
  # data3 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_1_5.mp4.npz', allow_pickle=True)
  # sample_2d_data_path = os.path.join(gestures_2d_path, class_label, f"{class_label}_{video_id}.mp4.npz")
  # data = np.load(sample_2d_data_path, allow_pickle=True)
  print(transpose_data(data1).shape)
  data1=transpose_data(data1)
  print(transpose_data(data2).shape)
  data2=transpose_data(data2)
  # transpose_data(data2)

  # data1=polar_angle_velocity(label1,f'{vid1}',63)
  # data2=polar_angle_velocity(label2,f'{vid2}',63)
  # print(data1.shape)
  # print(data2.shape)
  data = np.concatenate((data1, data2))

  print("Shape ",data.shape)
  min_indices=get_window(data)
  min_indices = rms(min_indices)
  print(min_indices)
  data=polar_velocity_acceleration_(data)
  t=pad_(data[:min_indices[0]+1],63)[0:63,5:11 ,3:]
  t=np.array([t])
  t=t.reshape(t.shape[0], 63, 6)
  p1=encoding[np.argmax(model.predict(t))]

  t1=pad_(data[min_indices[0]+3:],63)[0:63,5:11 ,3:]
  t1=np.array([t1])
  t1=t1.reshape(t1.shape[0], 63, 6)
  p2=encoding[np.argmax(model.predict(t1))]

  print(f"label vs pred 1-> {label1==p1} for :{label1},{p1}")
  print(f"label vs pred 2-> {label2==p2} for :{label2},{p2}")
  return label1==p1 and label2==p2
test2(model,"stop","warn","3_1","4_3")


(30, 17, 4)
(27, 17, 4)
Shape  (57, 17, 4)
[28, 54]
1/1 [==============================] - 0s 50ms/step
label vs pred 1-> False for :stop,no
label vs pred 2-> False for :warn,hello


False

In [ ]:
import itertools

gestures = ['stop', 'no', 'abort', 'circle', 'turn', 'turn_left', 'turn_right', 'warn',  'hello']
# numcombinations = ['1_{}'.format(i) for i in range(1, 6)]
numcombinations = ['1_3', '2_2', '3_5', '4_7', '3_1']

def generate_combinations(gestures, num_combinations):
    for gesture_pair in itertools.combinations(gestures, 2):
        for num_pair in itertools.permutations(num_combinations, 2):
            yield list(gesture_pair) + list(num_pair)

generator = generate_combinations(gestures, numcombinations)
t=[]
# Print first 10 combinations for demo
try:
  for i in range(1000):
    a,b,c,d=next(generator)
    t.append(test2(model,a, b, c, d))
except:
  print(t.count(True),len(t),(t.count(True)/len(t))*100)




Streaming output truncated to the last 5000 lines.
(27, 17, 4)
Shape  (53, 17, 4)
[25, 49]
1/1 [==============================] - 0s 68ms/step
label vs pred 1-> True for :stop,stop
label vs pred 2-> False for :turn_left,turn
(30, 17, 4)
(21, 17, 4)
Shape  (51, 17, 4)
[28, 50]
1/1 [==============================] - 0s 76ms/step
label vs pred 1-> False for :stop,no
label vs pred 2-> True for :turn_left,turn_left
(30, 17, 4)
(26, 17, 4)
Shape  (56, 17, 4)
[28, 53]
1/1 [==============================] - 0s 42ms/step
label vs pred 1-> False for :stop,no
label vs pred 2-> False for :turn_left,hello
(30, 17, 4)
(32, 17, 4)
Shape  (62, 17, 4)
[28, 61]
1/1 [==============================] - 0s 51ms/step
label vs pred 1-> False for :stop,no
label vs pred 2-> True for :turn_left,turn_left
(30, 17, 4)
(26, 17, 4)
Shape  (56, 17, 4)
[28, 55]
1/1 [==============================] - 0s 40ms/step
label vs pred 1-> False for :stop,no
label vs pred 2-> False for :turn_left,turn
(23, 17, 4)
(28, 17, 4)
Sh

In [ ]:
print(t.count(True),len(t),f"|| Accuracy :",(t.count(True)/len(t))*100)

307 720 || Accuracy : 42.638888888888886


In [ ]:
print(t.count(True),len(t),(t.count(True)/len(t))*100)

Accuracy :42.63

<a name="bonus-B2"></a>
# Bonus item B2
---

<a name="9-gestures"></a>
## Extend your work to training a model with all 9 gestures recognition for 3-gesture sequence recognition
---

In [ ]:
def test3(model,label1,label2,label3,vid1,vid2,vid3):
  # Load the data
  data1 = np.load(f'./drive/My Drive/CS256Project/data/gestures_basic_d2/{label1}/{label1}_{vid1}.mp4.npz', allow_pickle=True)['keypoints']
  data2 = np.load(f'./drive/My Drive/CS256Project/data/gestures_basic_d2/{label2}/{label2}_{vid2}.mp4.npz', allow_pickle=True)['keypoints']
  data3 = np.load(f'./drive/My Drive/CS256Project/data/gestures_basic_d2/{label2}/{label2}_{vid3}.mp4.npz', allow_pickle=True)['keypoints']
  # data3 = np.load('./drive/My Drive/CS256Project/data/gestures_basic_d2/no/no_1_5.mp4.npz', allow_pickle=True)
  # sample_2d_data_path = os.path.join(gestures_2d_path, class_label, f"{class_label}_{video_id}.mp4.npz")
  # data = np.load(sample_2d_data_path, allow_pickle=True)
  print(transpose_data(data1).shape)
  data1=transpose_data(data1)
  print(transpose_data(data2).shape)
  print(transpose_data(data3).shape)
  data2=transpose_data(data2)
  data3=transpose_data(data3)
  # transpose_data(data2)

  # data1=polar_angle_velocity(label1,f'{vid1}',63)
  # data2=polar_angle_velocity(label2,f'{vid2}',63)
  # print(data1.shape)
  # print(data2.shape)
  data = np.concatenate((data1, data2, data3))

  print("Shape ",data.shape)
  min_indices=get_window(data)
  min_indices = rms(min_indices)
  print(min_indices)
  data=polar_velocity_acceleration_(data)
  t=pad_(data[:min_indices[0]+1],63)[0:63,5:11 ,3:]
  t=np.array([t])
  t=t.reshape(t.shape[0], 63, 6)
  p1=encoding[np.argmax(model.predict(t))]

  t1=pad_(data[min_indices[0]+3:min_indices[1]],63)[0:63,5:11 ,3:]
  t1=np.array([t1])
  t1=t1.reshape(t1.shape[0], 63, 6)
  p2=encoding[np.argmax(model.predict(t1))]

  t2=pad_(data[min_indices[1]+1:],63)[0:63,5:11 ,3:]
  t2=np.array([t2])
  t2=t2.reshape(t2.shape[0], 63, 6)
  p3=encoding[np.argmax(model.predict(t2))]

  print(f"label vs pred 1-> {label1==p1} for :{label1},{p1}")
  print(f"label vs pred 2-> {label2==p2} for :{label2},{p2}")
  print(f"label vs pred 2-> {label3==p3} for :{label3},{p3}")
  return label1==p1 and label2==p2 and label3==p3
test3(model,"stop",'warn','abort','3_1','4_3','4_1')

(30, 17, 4)
(27, 17, 4)
(28, 17, 4)
Shape  (85, 17, 4)
[28, 81]
1/1 [==============================] - 0s 66ms/step
label vs pred 1-> False for :stop,no
label vs pred 2-> False for :warn,hello
label vs pred 2-> False for :abort,hello


False

In [ ]:
import itertools

gestures = ['stop', 'no', 'abort', 'circle', 'turn', 'turn_left', 'turn_right', 'warn',  'hello']
# numcombinations = ['1_{}'.format(i) for i in range(1, 6)]
numcombinations = ['1_3', '2_2', '3_5', '4_7', '3_1']

def generate_combinations(gestures, num_combinations):
    for gesture_pair in itertools.combinations(gestures, 3):
        for num_pair in itertools.permutations(num_combinations, 3):
            yield list(gesture_pair) + list(num_pair)

generator = generate_combinations(gestures, numcombinations)
t=[]
# Print first 10 combinations for demo
try:
  for i in range(500):
    a,b,c,d,e,f=next(generator)
    try:
      t.append(test3(model,a, b, c, d, e, f))
    except:
      continue
except:
  print(t.count(True),len(t),(t.count(True)/len(t))*100)

Streaming output truncated to the last 5000 lines.
label vs pred 1-> True for :stop,stop
label vs pred 2-> True for :no,no
label vs pred 2-> False for :abort,hello
(35, 17, 4)
(35, 17, 4)
(44, 17, 4)
Shape  (114, 17, 4)
[70, 110]
1/1 [==============================] - 0s 44ms/step
label vs pred 1-> False for :stop,hello
label vs pred 2-> False for :no,circle
label vs pred 2-> False for :abort,stop
(35, 17, 4)
(24, 17, 4)
(22, 17, 4)
Shape  (81, 17, 4)
[34, 61]
1/1 [==============================] - 0s 44ms/step
label vs pred 1-> True for :stop,stop
label vs pred 2-> False for :no,hello
label vs pred 2-> False for :abort,hello
(35, 17, 4)
(24, 17, 4)
(35, 17, 4)
Shape  (94, 17, 4)
[34, 92]
1/1 [==============================] - 0s 43ms/step
label vs pred 1-> True for :stop,stop
label vs pred 2-> True for :no,no
label vs pred 2-> False for :abort,stop
(35, 17, 4)
(24, 17, 4)
(44, 17, 4)
Shape  (103, 17, 4)
[59, 99]
1/1 [==============================] - 0s 47ms/step
label vs pred 1-> Fal

In [ ]:
print(t.count(True),len(t),f"|| Accuracy",(t.count(True)/len(t))*100)

159 452 || Accuracy 35.176991150442475


<a name="additional_results"></a>
# Additional related results not shown above
---